### PR TITLE
Add guards for mcore imports [GPT FIM]

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_fim_dataset.py [hunk 1]
```diff
@@ -15,10 +15,21 @@
...
+    from megatron.core.datasets.indexed_dataset import IndexedDataset
+    from megatron.core.datasets.utils import Split
+
+    HAVE_MEGATRON_CORE = True
+
+except (ImportError, ModuleNotFoundError):
+
+    GPTDataset = GPTDatasetConfig = IndexedDataset = Split = ApexGuardDefaults
+
+    HAVE_MEGATRON_CORE = False
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_fim_dataset.py:32]: Need to capture the error here

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_fim_dataset.py [hunk 2]
```diff
@@ -33,6 +44,11 @@ class GPTFIMDatasetConfig(GPTDatasetConfig):
     """
 
     def __init__(self, fim, **kwargs):
+        if not HAVE_MEGATRON_CORE:
+            raise ImportError(
+                "megatron-core was not found. Please see the NeMo README for installation instructions: https://github.com/NVIDIA/NeMo#megatron-gpt."
+            )
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_fim_dataset.py:50]: Need to log the captured error from above here

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_fim_dataset.py [hunk 3]
```diff
@@ -62,6 +78,11 @@ def __init__(
         index_split: Split,
         config: GPTFIMDatasetConfig,
     ) -> None:
+        if not HAVE_MEGATRON_CORE:
+            raise ImportError(
+                "megatron-core was not found. Please see the NeMo README for installation instructions: https://github.com/NVIDIA/NeMo#megatron-gpt."
+            )
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_fim_dataset.py:84]: Need to log the captured error from above here

**Comment by dimapihtar** [comment]
jenkins

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

We should go through and update all the import guards to capture the error and log it

**Pull request merged**